pest/src/span.rs

```diff
@@ -232,29 +256,47 @@ impl<'i> Hash for Span<'i> {
     }
 }
 
-/// Line iterator for Spans, created by [`Span::lines()`].
+/// Line iterator for Spans, created by [`Span::lines_span()`].
 ///
-/// Iterates all lines that are at least partially covered by the span.
+/// Iterates all lines that are at least _partially_ covered by the span. Yielding a `Span` for each.
 ///
-/// [`Span::lines()`]: struct.Span.html#method.lines
-pub struct Lines<'i> {
+/// [`Span::lines_span()`]: struct.Span.html#method.lines_span
+pub struct LinesSpan<'i> {
     span: &'i Span<'i>,
     pos: usize,
 }
 
-impl<'i> Iterator for Lines<'i> {
-    type Item = &'i str;
-    fn next(&mut self) -> Option<&'i str> {
+impl<'i> Iterator for LinesSpan<'i> {
+    type Item = Span<'i>;
+    fn next(&mut self) -> Option<Self::Item> {
         if self.pos > self.span.end {
             return None;
         }
         let pos = position::Position::new(self.span.input, self.pos)?;
         if pos.at_end() {
             return None;
         }
-        let line = pos.line_of();
+
+        let line_start = pos.find_line_start();
         self.pos = pos.find_line_end();
-        Some(line)
+
+        Span::new(self.span.input, line_start, self.pos)
```

Comment:
In the previous implementation this unsafely used `Span::new_unchecked`. It _should_ be safe because `find_line_end` should never extend pass `input`. We could save a few bounds check however it's not necessary.